swap/src/seed.rs

```diff
@@ -41,6 +43,19 @@ impl Seed {
         Ok(private_key)
     }
 
+    /// Returns a non-hardened xpub derived from a xprv which in return is
+    /// derived from the seed
+    pub fn derive_extended_public_key(&self, network: bitcoin::Network) -> Result<ExtendedPubKey> {
+        let priv_key = self.derive_extended_private_key(network)?;
+        let secp = Secp256k1::new();
+
+        // we take the first child private key to derive a non_hardened xpub key
+        let non_hardened_xprv = priv_key.ckd_priv(&secp, ChildNumber::from_normal_idx(0)?)?;
+
+        let public_key = ExtendedPubKey::from_private(&secp, &non_hardened_xprv);
+        Ok(public_key)
+    }
```

Comment:
So we are able to spend coins sent to this address because it is a child-key of the xprv that we use for the other wallet?

This is not obvious from my point of view. Can we just use the internal wallet for everything if the user didn't specify an xpub instead of deriving one that we happen to know the xprv for?

I am worried that the invariants that need to be maintained here for things to work are too complicated and we end up deriving addresses that we don't actually have the keys for.

Reply:
Yes, we could also just pass around the _same_ wallet twice, e.g. either clone or instantiate it again from `bitcoin::Wallet`. Imho this is harder to understand for a developer than having two keys, one for each wallet. 

> I am worried that the invariants that need to be maintained here for things to work are too complicated and we end up deriving addresses that we don't actually have the keys for.

I don't follow your concern: if we derive an xpub from the user's seed we should always be able to also derive the private key from it. 
If I misunderstood you, can you rephrase please?

Reply:
> I don't follow your concern: if we derive an xpub from the user's seed we should always be able to also derive the private key from it.

Yes, as long as the rest of the application follows along here.
What I am saying is, for this to work, several parts of the code need to be kept semantically in sync and that part is not obvious!

To demonstrate what I mean: The "internal" wallet will find coins on the chain that it never handed out an address for. It just happened that a different piece of code, outside the wallet, derived an xpub that belongs to the same seed. Hence, the "ownership" of the xpriv within the wallet is broken.

Having to keep this in mind when modifying the code creates unnecessary complexity.

> Imho this is harder to understand for a developer than having two keys, one for each wallet.

I don't follow how this is harder to understand?
If you trace the code and you see the following:

```rust
let internal_wallet = // ...;

let personal_wallet = if xpub { new_personal_wallet(xpub) } else { personal_wallet.clone() };
```

All you need to keep in your head is:

"So if no xpub is set, I can assume the two wallets are the same."

No Bitcoin specific knowledge required about how derivation paths work etc.
You can simply discard the additional information about two wallets existing and conceptually "merge" them in your head.

I'll claim that it is easier to understand what an application does if you can follow the dataflow of its variables than if you need to have domain-specific knowledge to work out that addresses handed out be one wallet will end up in the balance of the other one. That is just counter-intuitive.

Reply:
> What I am saying is, for this to work, several parts of the code need to be kept semantically in sync and that part is not obvious!

There is the seed generation and the wallet instantiation. I don't see other moving parts which need to keep in sync. 

> "So if no xpub is set, I can assume the two wallets are the same."

I might have misunderstood your initial comment when I answered with : 
> Yes, we could also just pass around the same wallet twice, e.g. either clone or instantiate it again from bitcoin::Wallet. Imho this is harder to understand for a developer than having two keys, one for each wallet.

What I wanted to say with this that I find it harder to understand what you pointed out (We know we have different understandings about code complexity so I won't argue about this any further. :) 
):  

> "So if no xpub is set, I can assume the two wallets are the same."

I'm coming to a different conclusion: the code here is wrong. 
The two wallets should not be the same. 
We should not re-use the same xprv-key to derive the xpub for the personal wallet as there should not be a relation between them. We should just derive a different xprv-key from the seed for the personal wallet.

Reply:
>  The two wallets should not be the same.

But at the moment (without this feature) they are, right? Is your claim that the current code and behaviour is wrong?

My goal would have been: don't "fix" something that is not broken. If we simply clone the original wallet, the xpub functionality is layered on top of what we currently have, without any changes in functionality if it is disabled. Given that we seem to be happy with how the code currently works, I don't see a reason to not do that.

> We should not re-use the same xprv-key to derive the xpub for the personal wallet as there should not be a relation between them. We should just derive a different xprv-key from the seed for the personal wallet.

That would address my concerns about "ownership" of the keys with each wallet. However, I think we care getting off-track here with the feature design in terms of what the original motivation is. I am not sure the concept of a personal vs internal wallet overall makes a whole lot of sense.

1. At the moment, the ASB doesn't need a Bitcoin wallet for anything other than receiving money (the other protocol is still in the works). Hence, even if you want to keep two wallets, there shouldn't be a need to pass a personal-wallet into any Alice / ASB related code.
2. Once we allow trading in both directions, Alice needs to be able to transfer BTC coins into a protocol as well. For that, it will need an xprv to sign transactions (i.e. an xpub is not enough).

If we have a personal and an internal wallet and the xpub feature is not used, which one should be used for (2)? If it is the personal wallet, what job does the internal wallet have? If it is the internal wallet, why can't we reuse the funds from the personal one for swaps in the other direction?

It appears to me that the ASB only has a single wallet and that is an internal one (internal because we control the keys). The only thing we want is to be able to directly send money out of the ASB without an intermediary transaction through the internal wallet. That is an _optimization_ of existing functionality, not a new, orthogonal feature. However, the current code doesn't make it very clear that we are optimization something. Rather, it suggests that there are reasons why we don't to manage to different sets of keys and I don't see why we should be doing that.

Keep in mind that we sort of have two wallets in the code: We have the ephemeral keys that we use for the protocol and an actual wallet that is used to hold all funds that we receive eventually. s Is that what you are trying to express with "internal" and "personal" wallet?

We don't use the wallet to derive these keys so as expressed above, we don't actually need a wallet instance for Alice at the moment for anything other than getting the two final addresses.

Reply:
> But at the moment (without this feature) they are, right? Is your claim that the current code and behaviour is wrong?
> [...]
> Keep in mind that we sort of have two wallets in the code: We have the ephemeral keys that we use for the protocol and an actual wallet that is used to hold all funds that we receive eventually. Is that what you are trying to express with "internal" and "personal" wallet?

Wrong might be a strong word but yes, I think the we should have two different wallets if the 'internal' wallet even qualifies to be called as a wallet as it is only used to sign protocol transactions. 

Anyway, changing this is out of scope of this ticket. 

What I wanted to say with: 

> I'm coming to a different conclusion: the code here is wrong.
The two wallets should not be the same.

Is that the from me newly introduced code is wrong. We should take the chance to decouple the two wallets so that there is no relation between them and not derive the xpub from the same xprv as used for the internal wallet. 

This is also useful when the ASB can trade in both directions, i.e. it has one wallet that holds the money and one _wallet_ which is used to sign protocol transactions.

Reply:
> This is also useful when the ASB can trade in both directions, i.e. it has one wallet that holds the money and one wallet which is used to sign protocol transactions.

Okay, I think we finally came to the bottom of all this!

We don't use `bitcoin::Wallet` to sign protocol transactions. Never. All protocol transactions are signed with very specific private keys that are generated completely randomly during execution setup.

The only transactions signed by `bitcoin::Wallet` are:

- Funding `TxLock`, which by itself does not qualify as a protocol transaction as it spends arbitrary inputs
- Withdrawing to an address

All other transactions:

- spending TxLock as redeem
- spending TxLock as cancel
- spending TxCancel as punish
- spending TxCancel as refund

are not signed by `bitcoin::Wallet` but by randomly generated keys.

As such, I don't see why we need **two** `bitcoin::Wallet` instances.

Reply:
riiight... I don't know what I thought and followed blindly the recommendation in the issue...

Since this is the case, we definitely don't need two wallets. In fact, this ticket is way simpler than this PR proposes and can be reduced to very few changes. 

We let the user define a xpub key (or *pub as defined in [SLIP-132](https://github.com/satoshilabs/slips/blob/master/slip-0132.md)) and instantiate the wallet from this key. If it is an xpub key we have a read-only wallet but this is fine if we only swap in one one direction. 
If no key was provided, we derive one from the internal seed as we used to.

Reply:
cc @da-kami :⬆️ this conclusion is crucial if you take over this PR.